File: Algorithms/Sorting/MergeSort/Pascal/merge-sort.pas
(**
 *  Author: Nhat M. Nguyen
 *  Date: 05-02-18
**)

program merge_sort;


uses crt;


const
  MAXN = 100000;
  INF = 100000;


var
  i, n: longint;
  a: array [0..MAXN] of longint;


procedure merge(var a: array of longint; st: longint; mi: longint; en: longint);

  var
    i, j, k: longint;
    n_left, n_right: longint;
    left: array of longint;
    right: array of longint;

  begin
    n_left := mi - st + 1;
    n_right := en - (mi + 1) + 1;
    
    setlength(left, n_left + 1);
    setlength(right, n_right + 1);

    for i := st to mi do
      left[i - st] := a[i];
    
    for j := mi + 1 to en do
      right[j - mi - 1] := a[j];
    
    left[n_left] := INF;
    right[n_right] := INF;
    i := 0;
    j := 0;

    for k := st to en do
      if left[i] < right[j] then
        begin
          a[k] := left[i];
          i := i + 1;
        end
      else
        begin
          a[k] := right[j];
          j := j + 1;
        end;
  end;


procedure merge_sort(var a: array of longint; st: longint; en: longint);

  var
    mi: longint;

  begin
    if st < en then
      begin
        mi := (st + en) div 2;
        merge_sort(a, st, mi);
        merge_sort(a, mi + 1, en);
        merge(a, st, mi, en);
      end;
  end;


begin
  clrscr;
  readln(n);
  for i := 0 to n - 1 do
    readln(a[i]);
  merge_sort(a, 0, n - 1);
  for i := 0 to n - 1 do
    writeln(a[i]);
end.

File: Algorithms/Sorting/MergeSort/Java/MergeSort.java
/**
 * @author Nhat M. Nguyen
 * @date   25-02-18
**/


class MergeSort {
    /**
     * @param a     array
     * @param start index of the start position, inclusive
     * @param end   index of the end position, inclusive
     */
    public void sort(int[] a, int start, int end) {
        // If the length of the subarray is 1, it is already sorted.
        if (start < end) {
            int mid = (start + end) / 2;
            sort(a, start, mid);
            sort(a, mid + 1, end);
            merge(a, start, mid, end);
        }
    }

    /**
     * @param a     array
     * @param start index of the start position, inclusive
     * @param mid   index of the mid position;
     *              a[mid] belongs to the left part
     * @param end   index of the end position, inclusive
     */
    private void merge(int[] a, int start, int mid, int end) {
        // Create copies of the left and right subarrays
        int nLeft = mid - start + 1;
        int nRight = end - (mid + 1) + 1;
        int[] left = new int[nLeft + 1];
        int[] right = new int[nRight + 1];

        for (int i = 0; i < nLeft; i++) {
            left[i] = a[start + i];
        }
        for (int i = 0; i < nRight; i++) {
            right[i] = a[mid + 1 + i];
        }

        // Small trick: last element of each subarray is set to infinity
        left[nLeft] = Integer.MAX_VALUE;
        right[nRight] = Integer.MAX_VALUE;

        int i = 0;
        int j = 0;

        for (int k = start; k <= end; k++) {
            if (left[i] < right[j]) {
                a[k] = left[i];
                i++;
            }
            else {
                a[k] = right[j];
                j++;
            }
        }
    }
}
